File: packages/connector/services/toggl_track/orchestrator.py
"""Toggl Track オーケストレーター

全データ同期の統合エントリーポイント。
マスタ → エントリーの順で実行（マスタ内は並列）。
"""

import time
from typing import TypedDict

from pipelines.lib.logger import setup_logger
from pipelines.services.toggl_track.sync_time_entries import sync_time_entries
from pipelines.services.toggl_track.sync_masters import sync_masters

logger = setup_logger(__name__)


class SyncAllResult(TypedDict):
    """全同期結果"""
    success: bool
    time_entries_count: int
    masters_counts: dict[str, int]
    elapsed_seconds: float


async def sync_all(days: int = 3) -> SyncAllResult:
    """Toggl Track の全データを同期

    マスタ → エントリーの順で実行。
    マスタ内は並列で取得。

    Args:
        days: 同期する日数（time_entries用）

    Returns:
        同期結果
    """
    start_time = time.perf_counter()
    logger.info(f"Starting Toggl Track full sync ({days} days)")

    errors = []
    time_entries_count = 0
    masters_counts = {}

    try:
        # 1. マスタ同期（内部で並列実行）
        logger.info("Step 1: Syncing masters...")
        masters_result = await sync_masters()
        masters_counts = masters_result["counts"]

        if not masters_result["success"]:
            errors.append("masters: partial failure")
            logger.warning("Masters sync had partial failures, continuing with time entries...")

        # 2. エントリー同期
        logger.info("Step 2: Syncing time entries...")
        entries_result = await sync_time_entries(days=days)
        time_entries_count = entries_result["count"]

        elapsed = round(time.perf_counter() - start_time, 2)

        # ログ出力
        masters_str = ", ".join(f"{k}={v}" for k, v in masters_counts.items())
        logger.info(
            f"Toggl Track full sync completed in {elapsed}s: "
            f"{masters_str}, time_entries={time_entries_count}"
        )

        if errors:
            logger.warning(f"Some syncs had issues: {errors}")

        return SyncAllResult(
            success=len(errors) == 0,
            time_entries_count=time_entries_count,
            masters_counts=masters_counts,
            elapsed_seconds=elapsed,
        )

    except Exception as e:
        elapsed = round(time.perf_counter() - start_time, 2)
        logger.error(f"Toggl Track full sync failed after {elapsed}s: {e}")
        raise

File: packages/connector/src/services/toggl-track/orchestrator.ts
/**
 * Toggl Track Orchestrator
 *
 * Unified entry point for full data sync.
 * Executes master sync first, then time entries.
 */

import { setupLogger } from "../../lib/logger.js";
import { syncTimeEntries } from "./sync-time-entries.js";
import { syncMasters } from "./sync-masters.js";

const logger = setupLogger("toggl-orchestrator");

// Types
export interface SyncAllResult {
  success: boolean;
  timeEntriesCount: number;
  mastersCounts: Record<string, number>;
  elapsedSeconds: number;
}

/**
 * Sync all Toggl Track data
 *
 * Executes in order: masters -> time entries
 *
 * @param days - Number of days for time entries
 * @returns Sync result
 */
export async function syncAll(days: number = 3): Promise<SyncAllResult> {
  const startTime = performance.now();
  logger.info(`Starting Toggl Track full sync (${days} days)`);

  const errors: string[] = [];
  let timeEntriesCount = 0;
  let mastersCounts: Record<string, number> = {};

  try {
    // 1. Masters sync (parallel internally)
    logger.info("Step 1: Syncing masters...");
    const mastersResult = await syncMasters();
    mastersCounts = mastersResult.counts;

    if (!mastersResult.success) {
      errors.push("masters: partial failure");
      logger.warn("Masters sync had partial failures, continuing with time entries...");
    }

    // 2. Time entries sync
    logger.info("Step 2: Syncing time entries...");
    const entriesResult = await syncTimeEntries(days);
    timeEntriesCount = entriesResult.count;

    const elapsed = Math.round((performance.now() - startTime) / 10) / 100;

    // Log summary
    const mastersStr = Object.entries(mastersCounts)
      .map(([k, v]) => `${k}=${v}`)
      .join(", ");
    logger.info(
      `Toggl Track full sync completed in ${elapsed}s: ${mastersStr}, time_entries=${timeEntriesCount}`
    );

    if (errors.length > 0) {
      logger.warn(`Some syncs had issues: ${errors.join(", ")}`);
    }

    return {
      success: errors.length === 0,
      timeEntriesCount,
      mastersCounts,
      elapsedSeconds: elapsed,
    };
  } catch (error) {
    const elapsed = Math.round((performance.now() - startTime) / 10) / 100;
    logger.error(`Toggl Track full sync failed after ${elapsed}s: ${error}`);
    throw error;
  }
}
